ui/raidboss/emulator/data/NetworkLogConverter.ts

```diff
@@ -0,0 +1,41 @@
+import EventBus from '../EventBus';
+import LineEvent from './network_log_converter/LineEvent';
+import LogRepository from './network_log_converter/LogRepository';
+import ParseLine from './network_log_converter/ParseLine';
+
+export default class NetworkLogConverter extends EventBus {
+  convertFile(data: string): LineEvent[] {
+    const repo = new LogRepository();
+    return this.convertLines(
+        // Split data into an array of separate lines, removing any blank lines.
+        data.split(NetworkLogConverter.lineSplitRegex).filter((l) => l !== ''),
+        repo,
+    );
+  }
+
+  convertLines(lines: (string | LineEvent)[], repo: LogRepository): LineEvent[] {
+    repo = repo || new LogRepository();
+    // This is a bit ugly but we can't keep the original lines array values around
+    // due to memory issues
+    // Convert from string to LineEvent
+    lines =
+      (
+        (lines as string[])
+          .map((l) => ParseLine.parse(repo, l)) as (LineEvent | boolean)[])
+        .filter((l) => l) as LineEvent[];
+
+    // Call `convert` to convert the network line to non-network format and update indexing values
+    lines = (lines as LineEvent[])
+      .map((l, i) => {
+        l.convert(repo);
+        l.index = i;
+        return l;
+      });
+
+    // Sort the lines based on `${timestamp}_${index}` to handle out-of-order lines properly
+    return (lines as LineEvent[])
+      .sort((l, r) => (`${l.timestamp}_${l.index}`).localeCompare(`${r.timestamp}_${r.index}`));
```

Comment:
There is...a lot going on in this function.  Can I suggest something that doesn't require `as` everywhere?

```suggestion
   const lineEvents = lines.map((l) => ParseLine.parse(repo, l)).filter(isLineEvent);

    // Call `convert` to convert the network line to non-network format and update indexing values
    lineEvents.map((l, i) => {
      l.convert(repo);
      l.index = i;
      return l;
    });

    // Sort the lines based on `${timestamp}_${index}` to handle out-of-order lines properly
    return lineEvents.sort((l, r) => (`${l.timestamp}_${l.index}`).localeCompare(`${r.timestamp}_${r.index}`));
```

Plus a type-guard function up above to filter undefined out without an `as`:

```typescript
const isLineEvent = (line?: LineEvent): line is LineEvent => {
  return !!line;
};
```

I'm not sure that I understand what the memory issues are here.  The caller is holding onto `lines` as well, so clobbering `lines` here doesn't remove the final ref.

Reply:
The issue isn't the strings in the array, the issue is the array itself.

Before the change to use the worker, the `lines` array passed into the function could have had an exceedingly large amount of elements, for instance the biggest single log file I have sitting around right now is 3,833,046 lines long.

That array alone is a little under 30mb just to store and track the references to the strings, not including the string data itself, and that's all assuming the javascript implementation is storing the array internally as an array of pointers and not as some more complex data type.

It's probably not nearly as important anymore and the adjustments you've suggested above would probably be fine, but it's potentially not and I'd prefer that someone double check the logic in worker to make sure I'm not missing some edge case where it might pass through a ton of lines, because that was the cause of a lot of browser crashes before.

Reply:
I think I just don't understand how changes inside of `convertLines` saves memory, as all callers of `convertLines` are still holding onto a ref to `lines`.  (Maybe the `convertFile` one doesn't because it's just a param, vs a named variable?  But that's a v8 detail I don't know enough about to say either way.)  But for the worker, the worker still has a ref on the array, so it doesn't matter if you clobber it inside of `convertLines` because there's still a ref on it and it can't be gc'd until the function is done and you set `lines = []` in the worker.  That's really the root of my confusion, as I don't understand how the previous code made any difference to the gc situation.

(Also, maybe this is not the PR to change it, but I still do not think that you should sort the lines.  The lines have out of order times in them but the order in the file is the order in which they are presented to the plugin.  So, I think it's incorrect to change that order.)

Reply:
In current state I don't think it's going to be a memory issue anymore. I'll make these adjustments.

As for sorting, the only reason that's in place is to address issues with CombatantTracker updating states out of order. I think that's a fix for after CombatantTracker and dependents have been converted to Typescript.

However, during testing of this I've noticed that there's an issue with the getter for `LineEvent.timestamp` that needs fixed. I'm going to get a PR in for that real quick before continuing on this PR.